src/scheduler.rs

```diff
@@ -0,0 +1,1013 @@
+        <T as Resource>::DynamicType: Default,
+    {
+        if self.modified {
+            return match self
+                .client
+                .merge_patch_status(resource, &json!({ "history": self.history }))
+                .await
+            {
+                Ok(res) => {
+                    self.modified = false;
+                    Ok(res)
+                }
+                err => err,
+            };
+        }
+
+        Ok(resource.clone())
+    }
+}
+
+impl SchedulerState {
+    pub fn new(current_mapping: PodToNodeMapping, new_mapping: PodToNodeMapping) -> Self {
+        SchedulerState {
+            current_mapping,
+            new_mapping,
+        }
+    }
+
+    pub fn mapping(&self) -> PodToNodeMapping {
+        self.current_mapping.merge(&self.new_mapping)
+    }
+
+    pub fn new_mapping(&self) -> PodToNodeMapping {
+        self.new_mapping.clone()
+    }
+}
+
+impl TryFrom<String> for PodIdentity {
+    type Error = Error;
+    fn try_from(s: String) -> Result<Self, Error> {
+        let split = s.split(';').collect::<Vec<&str>>();
+        if split.len() != 5 {
+            return Err(Error::PodIdentityNotParseable { pod_id: s });
+        }
+        Ok(PodIdentity {
+            app: split[0].to_string(),
+            instance: split[1].to_string(),
+            role: split[2].to_string(),
+            group: split[3].to_string(),
+            id: split[4].to_string(),
+        })
+    }
+}
+
+impl From<PodIdentity> for String {
+    fn from(pod_id: PodIdentity) -> Self {
+        format!(
+            "{};{};{};{};{}",
+            pod_id.app, pod_id.instance, pod_id.role, pod_id.group, pod_id.id
+        )
+    }
+}
+
+impl PodToNodeMapping {
+    pub fn from(pods: &[Pod], id_label_name: Option<&str>) -> Self {
+        let mut pod_node_mapping = PodToNodeMapping::default();
+        for pod in pods {
+            let labels = &pod.metadata.labels;
+            let app = labels.get(labels::APP_NAME_LABEL);
+            let instance = labels.get(labels::APP_INSTANCE_LABEL);
+            let role = labels.get(labels::APP_COMPONENT_LABEL);
+            let group = labels.get(labels::APP_ROLE_GROUP_LABEL);
+            let id = id_label_name.and_then(|n| labels.get(n));
+            pod_node_mapping.insert(
+                PodIdentity {
+                    app: app.cloned().unwrap_or_default(),
+                    instance: instance.cloned().unwrap_or_default(),
+                    role: role.cloned().unwrap_or_default(),
+                    group: group.cloned().unwrap_or_default(),
+                    id: id.cloned().unwrap_or_default(),
+                },
+                NodeIdentity {
+                    name: pod
+                        .spec
+                        .as_ref()
+                        .unwrap()
+                        .node_name
+                        .as_ref()
+                        .unwrap()
+                        .to_string(),
+                },
```

Comment:
I know that should be safe, but can we still remove the pure `unwrap()` calls and throw an error?